### DISCUSSION TITLE
Teardown

### DISCUSSION THREAD

**Discussion by mdaif** [pytest-dev/pytest-xdist#683] [General]
Hi,
First, thank you so much for supporting this library !
Second, I am having an issue and I was wondering if you could help me address it /direct me in the right way.
I have a session-wide fixture that creates a DB, yield, and delete the created DB, something like
```
@pytest.fixture(scope='session')
def test_db(worker_id):
    test_db_name = f'catalog-test-concurrent-{worker_id}'
    conn = psycopg2.connect(
       # Connect to an existing DB
   )
    conn.set_isolation_level(ISOLATION_LEVEL_AUTOCOMMIT)
    cursor = conn.cursor()
   # Create a test DB for this worker.
   conn.close()

  yield

  # connect to an existing DB.  
 # DROP test DB
 # close connection
```
What I am trying to do is to create a DB per process when the test runs, and drop it when the tests are finished/ an error happens/ an assertion error is raised.

The problem is, it seems like the teardown section (what comes after the yield statement) isn't executed for all processes. The creation part (before yield) is done successfully, but whatever comes after yield is executed for some of the workers and for some not (when the tests don't pass, i.e an assertion error is raised).

What might I be doing wrong here ?

**Comment by graingert** [comment]
worker processes can get killed for a variety of reasons - I tend to do all resource acquisition and teardown in the leader.

I also try to make setup idempotent - eg it would drop the database if it was there

**Comment by mdaif** [comment]
hi @graingert  Thank you for your response ! 
I think cleaning up after a fixture is a very important part. Sure I can drop the database if it's there when the test runs, but I'd still end up with dead test DBs after the tests are finished. Of course I can live with that, but I feel like not having the part after `yield` statement always executed is a bug that should be addressed.

**Comment by graingert** [comment]
> hi @graingert Thank you for your response !
> I think cleaning up after a fixture is a very important part. Sure I can drop the database if it's there when the test runs, but I'd still end up with dead test DBs after the tests are finished. Of course I can live with that, but I feel like not having the part after `yield` statement always executed is a bug that should be addressed.

Oh definitely, but the only reason I can think of for your teardown not executing is if pytest is aborted uncleanly, which is somewhat common in xdist scenarios.

If you could create an MRE with where your teardown doesn't get executed I could look into it more thoroughly

**Comment by mdaif** [comment]
Hi @graingert .. thanks a lot for your reply.
Unfortunately, I cannot do that because it's my company's codebase. 
I also have to add that I did the same cleanup in both setup and teardown ... and now the teardown seems to be executed fine !

**Comment by graingert** [comment]
> Hi @graingert .. thanks a lot for your reply.
> Unfortunately, I cannot do that because it's my company's codebase.
> I also have to add that I did the same cleanup in both setup and teardown ... and now the teardown seems to be executed fine !

An MRE wouldn't contain any of your company's codebase

**Reply by mdaif** [reply]
sorry .. can you elaborate ? I thought you mean a code snippet of what I am trying to do.

**Reply by graingert** [reply]
a minimal reproducible example, that's just new code written to excercise this fault

**Reply by mdaif** [reply]
Hi @graingert so sorry for the late reply, was pretty busy. Thank you for the elaboration :-)
Maybe we can close this ... since I got it working (by running the teardown as part of the setup), I will not spending more time on that.
Again, thank you so much for your help :))
